dwave/plugins/sklearn/utilities.py

```diff
@@ -0,0 +1,223 @@
+
+import numpy as np
+import numpy.typing as npt
+
+__all__ = ["corrcoef", "cov", "dot_2d"]
+
+
+def corrcoef(x: npt.ArrayLike, *,
+             out: typing.Optional[np.ndarray] = None,
+             rowvar: bool = True,
+             copy: bool = True,
+             ) -> np.ndarray:
+    """A drop-in replacement for :func:`numpy.corrcoef`.
+
+    This method is modified to avoid unnecessary memory usage when working with
+    :class:`numpy.memmap` arrays.
+    It does not support the full range of arguments accepted by
+    :func:`numpy.corrcoef`.
+
+    Args:
+        x: See :func:`numpy.corrcoef`.
+
+        out: Output argument. This must be the exact kind that would be returned
+            if it was not used.
+
+        rowvar: See :func:`numpy.corrcoef`.
+
+        copy: If ``True``, ``x`` is not modified by this function.
+
+    Returns:
+        See :func:`numpy.corrcoef`.
+
+    """
+    c = cov(x, out=out, rowvar=rowvar, copy=copy)
+    try:
+        d = np.diag(c)
+    except ValueError:
+        # scalar covariance
+        # nan if incorrect value (nan, inf, 0), 1 otherwise
+        return c / c
+    stddev = np.sqrt(d.real)
+    c /= stddev[:, None]
+    c /= stddev[None, :]
+
+    # Clip real and imaginary parts to [-1, 1].  This does not guarantee
+    # abs(a[i,j]) <= 1 for complex arrays, but is the best we can do without
+    # excessive work.
+    np.clip(c.real, -1, 1, out=c.real)
+    if np.iscomplexobj(c):
+        np.clip(c.imag, -1, 1, out=c.imag)
+
+    return c
+
+
+def cov(m: npt.ArrayLike, *,
+        out: typing.Optional[np.ndarray] = None,
+        rowvar: bool = True,
+        copy: bool = True,
+        ) -> np.ndarray:
+    """A drop-in replacement for :func:`numpy.cov`.
+
+    This method is modified to avoid unnecessary memory usage when working with
+    :class:`numpy.memmap` arrays.
+    It does not support the full range of arguments accepted by
+    :func:`numpy.cov`.
+
+    Args:
+        m: See :func:`numpy.cov`.
+
+        out: Output argument. This must be the exact kind that would be returned
+            if it was not used.
+
+        rowvar: See :func:`numpy.cov`.
+
+        copy: If ``True``, ``x`` is not modified by this function.
+
+    Returns:
+        See :func:`numpy.cov`.
+
+    """
+    # we want to modify X, so if copy=True we make a copy and re-call
+    if copy:
+        if hasattr(m, "flush"):
+            # we could do a lot of fiddling here, but it's easier to just
+            # disallow this case and rely on the user making a modifiable
+            # X
+            raise ValueError("memmap arrays cannot be copied easily")
+
+        return cov(np.array(m), rowvar=rowvar, copy=False, out=out)
+
+    # handle array-like
+    if isinstance(m, np.memmap):
+        X = m
+    else:
+        X = np.atleast_2d(np.asarray(m, dtype=np.result_type(m, np.float64)))
+
+    if X.ndim != 2:
+        raise ValueError("X must have 2 dimensions")
+
+    if not rowvar and X.shape[0] != 1:
```

Comment:
What is the `X.shape[0] != 1` condition for?

Reply:
No idea, it got inherited from the NumPy code